src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -112,24 +96,23 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
                 buf.ReadInt32(); // We don't care about the lower bounds
             }
 
-            if (dimensions == 0)
-                return new TAnyElement[0];   // TODO: static instance
-
-            var result = Array.CreateInstance(typeof(TAnyElement), dimLengths);
-
+            // We can avoid boxing in this case
             if (dimensions == 1)
             {
-                var oneDimensional = (TAnyElement[])result;
+                var oneDimensional = new TRequestedElement[dimLengths[0]];
                 for (var i = 0; i < oneDimensional.Length; i++)
-                    oneDimensional[i] = await _elementHandler.ReadWithLength<TAnyElement>(buf, async);
+                    oneDimensional[i] = await ElementHandler.ReadWithLength<TRequestedElement>(buf, async);
                 return oneDimensional;
             }
 
-            // Multidimensional
+            var result = Array.CreateInstance(typeof(TRequestedElement), dimLengths);
+
+            // Multidimensional arrays
+            // We can't avoid boxing here
             var indices = new int[dimensions];
             while (true)
             {
-                var element = await _elementHandler.ReadWithLength<TAnyElement>(buf, async);
+                var element = await ElementHandler.ReadWithLength<TRequestedElement>(buf, async);
                 result.SetValue(element, indices);
```

Comment:
Are you interested in optimizing this in the second PR? There is not so hard, just allocate a default boxed value before the loop in the case of value type, and use `Unsafe.Unbox<T>` to get the reference to the value and replace it each iteration passing the same boxed value to `SetValue`.

Reply:
That's a good Idea but I think that it's more complicated than you think as `Unsafe.Unbox<T>` has a `struct` type constraint which would mean that we have to use expression trees once again.
TBH I'd leave this for a follow-up PR.

Reply:
Take a look at `NullableHandler` which uses some kind of magic to read and write nullable values (: By the way, you can use the same trick instead of code generation via expressions.

Reply:
Ok, I gave it a closer look and immediately ran into the following problem:

> CS8177 C# Async methods cannot have by-reference locals

I think that for now there are enough obstacles to justify postponing this for a later PR.